core/src/main/java/io/grpc/internal/GrpcUtil.java

```diff
@@ -784,16 +784,18 @@ static ClientStreamTracer newClientStreamTracer(
       streamTracer = streamTracerFactory.newClientStreamTracer(info, headers);
     } else {
       streamTracer = new ForwardingClientStreamTracer() {
-        final ClientStreamTracer noop = new ClientStreamTracer() {};
-        AtomicReference<ClientStreamTracer> delegate = new AtomicReference<>(noop);
+        final AtomicBoolean initialized = new AtomicBoolean();
+        volatile ClientStreamTracer delegate = new ClientStreamTracer() {};
 
         void maybeInit(StreamInfo info, Metadata headers) {
-          delegate.compareAndSet(noop, streamTracerFactory.newClientStreamTracer(info, headers));
+          if (initialized.compareAndSet(false, true)) {
```

Comment:
Delegate may still be noop when maybeInit() returns. I think you should just use a lock.

Reply:
Right, it's my mistake.